include/rmm/mr/device/cuda_async_memory_resource.hpp

```diff
@@ -66,6 +68,39 @@ class cuda_async_memory_resource final : public device_memory_resource {
     win32     = 0x2,              ///< Allows a Win32 NT handle to be used for exporting. (HANDLE)
     win32_kmt = 0x4  ///< Allows a Win32 KMT handle to be used for exporting. (D3DKMT_HANDLE)
   };
+  /**
+   * @brief Constructs a cuda_async_memory_resource with the optionally specified initial pool size
+   * and release threshold.
+   *
+   * If the pool size grows beyond the release threshold, unused memory held by the pool will be
+   * released at the next synchronization event.
+   *
+   * @throws rmm::logic_error if the CUDA version does not support `cudaMallocAsync`
+   *
+   * @param initial_pool_size Optional initial size in bytes of the pool. If no value is provided,
+   * initial pool size is half of the available GPU memory.
+   * @param release_threshold Optional release threshold size in bytes of the pool. If no value is
+   * provided, the release threshold is set to the total amount of memory on the current device.
+   * @param export_handle_type Optional `cudaMemAllocationHandleType` that allocations from this
+   * resource should support interprocess communication (IPC). Default is
+   * `cudaMemHandleTypeNone` for no IPC support.
+   */
+  // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+  template <class Optional,
+            cuda::std::enable_if_t<cuda::std::is_same_v<cuda::std::remove_cvref_t<Optional>,
+                                                        thrust::optional<std::size_t>>,
+                                   int> = 0>
+  [[deprecated("Use std::optional instead of thrust::optional.")]]  //
```

Comment:
suggestion: please search the RAPIDS codebase before adding deprecations because repos set -Werror so deprecations break the build. If there are usages of thrust::optional in RAPIDS we should change those before deprecating.

Reply:
I looked at usage of `cuda_async_memory_resource` [here](https://github.com/search?q=org%3Arapidsai%20cuda_async_memory_resource&type=code)

And all occurrences that pass in additional arguments are from cuDF and passing in `Jlong` so I believe we are fine there.

For `` there is actually one [usage ](https://github.com/search?q=org%3Arapidsai+pool_memory_resource&type=code)with explicit `thrust::optional` in [node](https://github.com/rapidsai/node/blob/384a090e85ada374b3e1d476ecec91005937b854/modules/rmm/src/memory_resource.cpp#L73-L76)

That one is interesting because it is already calling the deprecated constructor that takes two optionals. Do we need to change anything there immediately given that it already is using a deprecated API?

Reply:
That one slipped through the cracks. I didn't think to search the node repo, as it's not part of the combined RAPIDS devcontainer. We should probably fix the deprecated usage, or at least file an issue. CC @trxcllnt

Reply:
The node repo is on RAPIDS release 22.12, so you can ignore it. I'll update the call site if/when node updates to 24.04.